pkg/controller/controller.go

```diff
@@ -409,9 +413,116 @@ func (c *Controller) handleClusterUpdate(key string) error {
+	if err != nil {
+		return fmt.Errorf("error listing pods: %v", err)
+	}
+
+	// TODO(schallert): per-cluster m3admin clients
+	placement, err := c.placementClient.Get()
+	if err != nil {
+		return fmt.Errorf("error fetching active placement: %v", err)
+	}
+
+	c.logger.Info("found placement", zap.Int("currentPods", len(pods)), zap.Int("placementInsts", placement.NumInstances()))
+
+	if placement.NumInstances() < len(pods) {
+		for _, pod := range pods {
+			_, ok := placement.Instance(pod.Name)
+			if ok {
+				continue
+			}
+
+			c.logger.Info("found pod not in placement", zap.String("pod", pod.Name))
+			inst, err := k8sops.PlacementInstanceFromPod(cluster, pod)
+			if err != nil {
+				err := fmt.Errorf("error creating instance for pod %s", pod.Name)
+				c.logger.Error(err.Error())
+				return err
+			}
+
+			reason := fmt.Sprintf("adding pod %s to placement", pod.Name)
+			cluster, err = c.setStatusPodBootstrapping(cluster, corev1.ConditionTrue, "PodAdded", reason)
+			if err != nil {
+				err := fmt.Errorf("error setting pod bootstrapping status: %v", err)
+				c.logger.Error(err.Error())
+				return err
+			}
+
+			err = c.placementClient.Add(*inst)
+			if err != nil {
+				err := fmt.Errorf("error adding pod to placement: %s", pod.Name)
+				c.logger.Error(err.Error())
+				return err
+			}
+
+			c.logger.Info("added pod to placement", zap.String("pod", pod.Name))
+			return nil
+		}
+	}
+
+	// Determine if any sets aren't at their desired replica count. Maybe we can
+	// reuse the set objects from above but being paranoid for now.
+	childrenSets, err = c.getChildStatefulSets(cluster)
+	if err != nil {
+		return err
+	}
+
+	for _, set := range childrenSets {
+		zone, ok := set.Labels["isolation-group"]
+		if !ok {
+			return fmt.Errorf("statefulset %s has no isolation-group label", set.Name)
+		}
+
+		group, ok := myspec.IsolationGroups(isoGroups).GetByName(zone)
+		if !ok {
+			return fmt.Errorf("zone %s not found in cluster isoGroups %v", zone, isoGroups)
+		}
+
+		if set.Spec.Replicas == nil {
+			return fmt.Errorf("set %s has unset spec replica", set.Name)
+		}
+
+		desired := group.NumInstances
+		current := *set.Spec.Replicas
+
+		// TODO(schallert): modify once we handle removes
+		if current >= desired {
+			continue
+		}
+
+		c.logger.Info("need to expand set, desired < current",
+			zap.String("set", set.Name),
+			zap.Int32("desired", desired),
+			zap.Int32("current", current))
+
+		set.Spec.Replicas = pointer.Int32Ptr(current + 1)
+		set, err = c.kubeClient.AppsV1().StatefulSets(set.Namespace).Update(set)
+		if err != nil {
+			return fmt.Errorf("error updating statefulset %s: %v", set.Name, err)
+		}
+
+		return nil
+	}
+
+	placement, err = c.placementClient.Get()
+	if err != nil {
+		return fmt.Errorf("error fetching placement: %v", err)
+	}
+
+	// See if we need to clean up the pod bootstrapping status.
+	cluster, err = c.reconcileBootstrappingStatus(cluster, placement)
```

Comment:
If pods are bootstrapping should we just keep re-queuing the cluster till it's completed?

Reply:
My plan so far was to wait until we get a new update when one of the pods transitions to ready. I've found that since the workqueue is rate-limiting, if we re-enqueue too many times it will take a really long time after that object is healthy for it to be processed depending on how many times we re-enqueue

Reply:
Got and make sense. Thanks for explaining.